features/GEM003_Unique-representation-identifier.feature

```diff
@@ -2,17 +2,11 @@
 @GEM
 Feature: GEM003 - Unique Representation Identifier
 
-  # Scenario: No Shape Representation identifier shall be used twice within the product representation of an element 
-
-      # Given An IfcProductDefinitionShape
-      #   And The element has more than 1 instance(s) of IfcShapeRepresentation
-
-      #   Then Each instance of IfcShapeRepresentation has a unique value for the attribute RepresentationIdentifier
-
-  Scenario: No Shape Representation identifier shall be used twice within the product representation of an element 
+  Scenario: A Shape Representation identifier must not be used twie within the product representation of an IfcProduct element
 
       Given An IfcProduct
         And Its values for attribute Representation
         And Its values for attribute Representations
+        And Its values for attribute RepresentationIdentifier
```

Comment:
I still cannot wrap my head around:

> *its* values for attribute *Representations* or *RepresentationIdentifier*

given that an IfcProduct has neither of them. 

If it is meant as: go down the attributes tree and find the entity with the fitting attribute, it doesn't seem to be very clear what is being checked if the attribute name is non-unique.

---
Scenario: Imagine ensuring 3D geometry on an IfcProduct:

    ....
    And  Its values for attribute Dim 
    Then The values must be 3

The file would have a product like:
IfcProduct
- IfcProduct.Representation = IfcProductDefinitionShape
- IfcProduct.Representation.Representations[0] = IfcShapeRepresentation
- IfcProduct.Representation.Representations[0].Items[0] = IfcExtrudedAreaSolid with **SELF/IfcSolidModel.Dim = 3**
- IfcProduct.Representation.Representations[0].Items[0].SweptArea = IfcArbitraryClosedProfileDef
- IfcProduct.Representation.Representations[0].Items[0].SweptArea.OuterCurve = IfcCurve with **IfcCurve.Dim = 2** (see IfcArbitraryClosedProfileDef.WR1)
- IfcProduct.Representation.Representations[0].Items[0].ExtrudedDirection = IfcDirection with **IfcDirection.Dim = HIINDEX(DirectionRatios), which may be 2 or 3**

What is the expected behavior? Should the product fail or pass the rule?

Reply:
I understand your confusion. `And` is not meant to be read here as a boolean and, but more a sequential chaining, think of it as tree/graph traversal. So:

```
Given An IfcProduct                                                # IfcProduct[]
        And Its values for attribute Representation                # IfcProduct[].Representation
        And Its values for attribute Representations               # IfcProduct[].Representation.Representations[]
        And Its values for attribute RepresentationIdentifier      # IfcProduct[].Representation.Representations[].RepresentationIdentifier
```

Do you have better ideas? We're also thinking about referencing concept templates from gherkin.

Reply:
I see. *its* applies to the object named in the expression above it. Thank you for the explanation.

A question, though: how to write the rule if one wishes to check the values of two attributes of the same object? For example, 

- `ExtrudedDirection` and `SweptArea` from my example above?
- `RepresentationIdentifier` and `RepresentationType` from your example above?

Would one need to write two separate scenarios?

Reply:
> how to write the rule if one wishes to check the values of two attributes 

It would be possible to recover multiple "bindings" from the stack frames. See #22 for an exploration. We'd have to connect the stack frames with the step definitions to build a mapping of attribute name -> value. Or store it as a dictionary in the context to begin with. So we might have to modify the approach a bit, but we're not too far away.

We can also think of a way to allow for explicit naming:

```
Given   An IfcProduct
  And   Its values for attribute Representation
  And   Its values for attribute Representations
  And   Its values for attribute RepresentationIdentifier to be referenced as RepId
  And   Its values for attribute RepresentationType to be referenced as RepType
  
 Then   RepId should be equal to RepType
```

(silly example obviously).